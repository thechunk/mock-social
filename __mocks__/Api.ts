module.exports = {
    UsersApi: {
        getAll: () => Promise.resolve([])
    }
};