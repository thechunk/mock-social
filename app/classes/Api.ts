import Config from 'react-native-config';

enum ApiPath {
    Users = '/users',
    Albums = '/albums',
    Photos = '/photos'
}
enum ApiQueryParam {
    UserId = 'userId',
    AlbumId = 'albumId'
}
enum HttpMethod {
    Get = 'GET',
    Post = 'POST',
    Put = 'PUT',
    Patch = 'PATCH',
    Delete = 'DELETE'
}
enum MimeType {
    ApplicationJson = 'application/json'
}

const apiUrl = (baseUrl: string, path: ApiPath, query?: string): RequestInfo => {
    return `${baseUrl}${path}${query ? `?${query}` : ''}` as RequestInfo;
};

const apiRequest = <T extends any>(url: RequestInfo, method?: HttpMethod, body?: string) => ((): Promise<T> => (
    fetch(url, {
        method: method ? method : HttpMethod.Get,
        headers: {
            Accept: MimeType.ApplicationJson,
            'Content-Type': MimeType.ApplicationJson
        },
        body: body != null && body.length > 0 ? JSON.stringify(body) : null
    })
        .then((response): T => response.json() as any as T)
));
const apiBaseUrl = Config.API_BASE_URL;

class UsersApi implements IUsersApi {
    public getAll() {
        return apiRequest<Array<IUser>>(apiUrl(apiBaseUrl, ApiPath.Users))()
    }

    public getById(id: number) {
        return apiRequest<IUser>(`${apiUrl(apiBaseUrl, ApiPath.Users)}/${id}`)()
    }
}
const Users = new UsersApi();

class AlbumsApi implements IAlbumsApi {
    public getAllByUserId(id: number) {
        const query = `${ApiQueryParam.UserId}=${id}`;
        return apiRequest<Array<IAlbum>>(apiUrl(apiBaseUrl, ApiPath.Albums, query))()
    }
}
const Albums = new AlbumsApi();

class PhotosApi implements IPhotosApi {
    public getAllByAlbumId(id: number) {
        const query = `${ApiQueryParam.AlbumId}=${id}`;
        return apiRequest<Array<IPhoto>>(apiUrl(apiBaseUrl, ApiPath.Photos, query))()
    }
}
const Photos = new PhotosApi();

export {Users as UsersApi};
export {Albums as AlbumsApi};
export {Photos as PhotosApi};
