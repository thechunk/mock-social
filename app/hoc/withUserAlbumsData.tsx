import React, {ComponentType, FC, useEffect, useState} from 'react';
import {AlbumsApi} from '../classes/Api';

const withUserAlbumsData = <P extends IWithUserAlbumsDataProps>(Component: ComponentType<P>):
    FC<HocOptional<P, IWithUserAlbumsDataProps> & IWithUserAlbumsDataOptionalProps> =>
    (props: HocOptional<P, IWithUserAlbumsDataProps> & IWithUserAlbumsDataOptionalProps) => {
        const [albums, setAlbums] = useState<Array<IAlbum>>([]);
        useEffect(() => {
            if (props.id) {
                AlbumsApi.getAllByUserId(props.id).then(setAlbums);
            }
        }, [props.id]);

        return (
            <Component {...props as P & IWithUserAlbumsDataOptionalProps} albums={albums} />
        );
    };

export default withUserAlbumsData;
