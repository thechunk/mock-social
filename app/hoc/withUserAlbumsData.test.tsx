/**
 * @jest-environment jsdom
 */

import * as React from 'react';
import {mount, shallow} from 'enzyme';
import {View} from "react-native";
import withUserAlbumsData from "./withUserAlbumsData";
import {AlbumsApi} from '../classes/Api';
require('setimmediate');

describe('withUserAlbumsData', () => {
    it('wrapper renders correctly', () => {
        const Component = () => (<View/>);
        const Wrapped: any = withUserAlbumsData(Component);
        expect(shallow(<Wrapped id={1} />)).toMatchSnapshot();
    });

    it('calls api', (done) => {
        AlbumsApi.getAllByUserId = jest.fn().mockReturnValue(Promise.resolve());
        const Component = () => (<div />);
        const Wrapped: any = withUserAlbumsData(Component);
        mount(<Wrapped id={1} />);
        done();
        expect(AlbumsApi.getAllByUserId).toBeCalledTimes(1);
    });

    it('calls api, sets data and loader status on success', (done) => {
        AlbumsApi.getAllByUserId = jest.fn().mockReturnValue(Promise.resolve(['foo']));
        const Component = () => (<div />);
        const Wrapped: any = withUserAlbumsData(Component);
        let wrappedComponent = mount(<Wrapped id={1} />).childAt(0);
        expect(wrappedComponent.props().loading).toEqual(true);

        flushAwaitable(done).then(() => {
            expect(wrappedComponent.props().loading).toEqual(false);
            expect(wrappedComponent.props().albums).toEqual(['foo']);
        });
    });
});
