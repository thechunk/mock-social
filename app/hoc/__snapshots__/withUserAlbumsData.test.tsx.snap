// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUserAlbumsData wrapper renders correctly 1`] = `
<Component
  albums={Array []}
  id={1}
  loading={true}
/>
`;
